packages/drivers/odsp-driver/src/compactSnapshotParser.ts

```diff
@@ -116,50 +116,92 @@ function readTreeSection(node: NodeCore) {
-					continue;
-				}
-			}
+				case 6: {
+					// "name": <node name>
+					// "nodeType": 3
+					// "value": <blob id>
+					if (
+						treeNode.getMaybeString(2) === "nodeType" &&
+						treeNode.getMaybeString(4) === "value"
+					) {
+						snapshotTree.blobs[treeNode.getString(1)] = treeNode.getString(5);
+						continue;
+					}
 
-			// "name": <node name>
-			// "nodeType": 3
-			// "value": <blob id>
-			if (
-				length === 6 &&
-				treeNode.getMaybeString(2) === "nodeType" &&
-				treeNode.getMaybeString(4) === "value"
-			) {
-				snapshotTree.blobs[treeNode.getString(1)] = treeNode.getString(5);
-				continue;
-			}
+					// "name": <node name>
+					// "children": <blob id>
+					// groupId: <group name>
+					if (
+						treeNode.getMaybeString(2) === "children" &&
+						treeNode.getMaybeString(4) === "groupId"
+					) {
+						const result = readTreeSection(treeNode.getNode(3));
+						trees[treeNode.getString(1)] = result.snapshotTree;
+						slowTreeStructureCount += result.slowTreeStructureCount;
+						snapshotTree.groupId = treeNode.getMaybeString(5);
+						continue;
+					}
 
-			// "name": <node name>
-			// "unreferenced": true
-			// "children": <blob id>
-			if (
-				length === 6 &&
-				treeNode.getMaybeString(2) === "unreferenced" &&
-				treeNode.getMaybeString(4) === "children"
-			) {
-				const result = readTreeSection(treeNode.getNode(5));
-				trees[treeNode.getString(1)] = result.snapshotTree;
-				slowTreeStructureCount += result.slowTreeStructureCount;
-				assert(treeNode.getBool(3), 0x3db /* Unreferenced if present should be true */);
-				snapshotTree.unreferenced = true;
-				continue;
+					// "name": <node name>
+					// "unreferenced": true
+					// "children": <blob id>
+					if (
+						treeNode.getMaybeString(2) === "unreferenced" &&
+						treeNode.getMaybeString(4) === "children"
+					) {
+						const result = readTreeSection(treeNode.getNode(5));
+						trees[treeNode.getString(1)] = result.snapshotTree;
+						slowTreeStructureCount += result.slowTreeStructureCount;
+						assert(
+							treeNode.getBool(3),
+							0x3db /* Unreferenced if present should be true */,
+						);
+						snapshotTree.unreferenced = true;
+						continue;
+					}
+					break;
+				}
+				case 8: {
+					// "name": <node name>
+					// "unreferenced": true
+					// "children": <blob id>
+					// groupId: <group name>
+					if (
+						treeNode.getMaybeString(2) === "unreferenced" &&
+						treeNode.getMaybeString(4) === "children" &&
+						treeNode.getMaybeString(6) === "groupId"
+					) {
+						const result = readTreeSection(treeNode.getNode(5));
+						trees[treeNode.getString(1)] = result.snapshotTree;
+						slowTreeStructureCount += result.slowTreeStructureCount;
+						assert(treeNode.getBool(3), "Unreferenced if present should be true");
+						snapshotTree.unreferenced = true;
+						snapshotTree.groupId = treeNode.getMaybeString(7);
+						continue;
```

Comment:
I remember I used some sample snapshot to validate that this logic actually made a difference in terms of performance. With more cases it could be that the value of approach is diminishing, and we are better of having only "default" flow below. Or maybe we want to reduce number of special cases such that smaller number of cases cover 95% of iterations (and thus they are faster) but remaining 5% go through slower default flow (but because we do not have special case for them, the other 95% are faster).

Not sure, just mentioning that we should re-evaluate approach as we make it more complex - the old reasons for the logic might no longer apply.

Reply:
As most of the tress won't have a groupid, so I have moved the if with groupid in end in switch, plus moving from if to switch would also be helpful here. So, let's keep it here as the original PR says that it makes tree section parsing 1.5x faster and we added 1 more block in length 6 and length 8 block just have 1 if. 
So, let's keep this.